### DISCUSSION TITLE
Working on new HTTP + JSON Powermeter fetcher  - inputs welcome

### DISCUSSION THREAD

**Discussion by berni2288** [hoylabs/OpenDTU-OnBattery#141] [Ideas]
Some people said that they would like to have support for HTTP + Json Powermeter reading, because they can't use MQTT.

Examples:
- Volkszähler / Energy meter + Tasmota
- Other Power meters that provide the current power consumption via HTTP and Json

I just started implementing this, but I need 2 things for this:

**Wanted: HTTP request client library**

- Should support HTTPS/SSL because this is state of the art nowadays and not supporting SSL seems weird to me
- If possible it should support async requests, because the ESP32 has a watchdog and could probably restart device if the request is blocking too long.
- Should support ESP32
- The Arduino HTTPClient class does not support async it seems

**Possible candiates:**

- https://github.com/khoih-prog/AsyncHTTPRequest_Generic - can do both things, but for some reason this repo is archived, so I would not use this.
- AsyncWebSocket.h from me-no-dev / ESPAsyncWebServer - suggested by @helgeerbe gonna have a look at this.
- [esp_http_client - From espressif ](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_client.html) - looks promising

Does anyone know something good?

**Wanted: An ESP32 Json library with JsonPath support**

- Should support [JsonPath](https://goessner.net/articles/JsonPath/) or some similar feature where you can select a single value in a Json document by providing some kind of path in one string.**
- ArduinoJson can't do this
- I've found the project https://github.com/mobizt/FirebaseJson which can fetch a value from a path provided by a string

Any other inputs/ideas are also welcome!

Current Webview design:

![image](https://user-images.githubusercontent.com/1029328/227163667-e12ce499-8db7-4a9e-813c-eae9f97e7cc0.png)

**Comment by helgeerbe** [comment]
Why not using the AsyncWebSocketClient from AsyncWebsocketServer?

**Reply by berni2288** [reply]
Does it support HTTPS/SSL?

I also just found this: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_client.html
this also looks promising.

**Reply by berni2288** [reply]
But a websocket is not quite what I want. web sockets are a subset of the HTTP protocol.

**Reply by helgeerbe** [reply]
You are right. If I remember right, Tasmota does not support web sockets. So the only thing what you can do ist to poll in main loop Tasmota. There is no asynchronus connection and if I remember right, Tasmota is not supporting ssl by default. It is possible to compile Tasmota with ssl support. But I doubt, that is really used.

**Reply by berni2288** [reply]
It's not just about Tasmota, it should be a generic solution for other power meters too. Maybe some support https or are cloud based? who knows.
SSL connections are state of the art today and if not necessary I would not want to send my credentials unencrypted over a network, you never know who is listening.
I could live without asynchronous requests.
Thanks, I'll just evaluate all the available options.

**Comment by Adminius** [comment]
Some Ideas and information (in german):
https://github.com/lumapu/ahoy/issues/606#issuecomment-1438587485
https://github.com/lumapu/ahoy/issues/606#issuecomment-1438698966

I mean, we need 3 URLs and not only one.
Some power meters have splitted values for export and input power and both are positive but only one is >0 simultaneously

**Comment by berni2288** [comment]
I decided to go with the [arduino-esp32/HTTPClient](https://github.com/espressif/arduino-esp32/blob/master/libraries/HTTPClient/src/HTTPClient.h). It seems to be a common way to do HTTP requests, It supports http and https, but there is no support for async mode.

I actually tried the [esp_http_client](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_client.html) first and would have preferred it because it seems quite major, but there were too many problems with it: async mode is supported for HTTPS only, there were conflicts with the HTTP_ enum with AsyncWebServer that could only be fixed by doing a lot of changes in the AsyncWebServer library and there were crashes during reading the HTTP response. The HTTP_ enum conflict is not only esp_http_client's fault.
Note to the makers of [esp_http_client](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_client.html)  & [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer): polluting the global namespace is a really terrible idea!!!11

Since there will be no async requests, I'l implement a default timeout of 1 second, that can be customized. I think this should do it and safe the ESP32 from reboots when the Powermeter is not reachable.

For Json parsing by path, I'll try out https://github.com/mobizt/FirebaseJson .

**Comment by berni2288** [comment]
PR is open now https://github.com/helgeerbe/OpenDTU-OnBattery/pull/153
reviews are welcome.
